File: ensark-frontend/src/app/core/models/loan.models.ts
import { LoanStatus, RepaymentStatus } from './enums';

export interface GuarantorRequest {
  name: string;
  phone: string;
  address: string;
  nidNumber: string;
  relation: string;
}

export interface LoanApplicationRequest {
  accountId: number;
  principalAmount: number;
  annualInterestRate: number;
  tenureMonths: number;
  guarantor: GuarantorRequest;
}

export interface GuarantorResponse {
  id: number;
  name: string;
  phone: string;
  address: string;
  nidNumber: string;
  relation: string;
  photoPath: string;
}

export interface DocumentResponse {
  id: number;
  fileName: string;
  originalFileName: string;
  contentType: string;
  fileSize: number;
}

export interface LoanApplicationResponse {
  loanId: number;
  accountId: number;
  accountNumber: string;
  principalAmount: number;
  annualInterestRate: number;
  tenureMonths: number;
  emiAmount: number;
  totalPayable: number;
  outstandingBalance: number;
  disbursementCharge: number;
  status: LoanStatus;
  applicationDate: string;
  approvalDate: string;
  disbursementDate: string;
  nextDueDate: string;
  rejectionReason: string;
  disbursementTransactionRef: string;
  guarantors: GuarantorResponse[];
  documents: DocumentResponse[];
}

export interface LoanRepaymentResponse {
  id: number;
  loanId: number;
  installmentNumber: number;
  dueDate: string;
  principalComponent: number;
  interestComponent: number;
  emiAmount: number;
  remainingBalanceAfter: number;
  status: RepaymentStatus;
  paidDate: string;
  transactionRef: string;
}

export interface LoanScheduleResponse {
  repaymentId: number;
  installmentNumber: number;
  dueDate: string;
  principalComponent: number;
  interestComponent: number;
  emiAmount: number;
  remainingBalanceAfter: number;
  status: RepaymentStatus;
  paidDate: string;
  transactionRef: string;
}

File: ensark-flutter/lib/models/loan/loan_models.dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import '../enums.dart';

part 'loan_models.freezed.dart';
part 'loan_models.g.dart';

@freezed
class LoanApplicationRequest with _$LoanApplicationRequest {
  const factory LoanApplicationRequest({
    int? accountId,
    double? principalAmount,
    double? annualInterestRate,
    int? tenureMonths,
    GuarantorRequest? guarantor,
  }) = _LoanApplicationRequest;

  factory LoanApplicationRequest.fromJson(Map<String, dynamic> json) => _$LoanApplicationRequestFromJson(json);
}

@freezed
class GuarantorRequest with _$GuarantorRequest {
  const factory GuarantorRequest({
    String? name,
    String? phone,
    String? address,
    String? nidNumber,
    String? relation,
  }) = _GuarantorRequest;

  factory GuarantorRequest.fromJson(Map<String, dynamic> json) => _$GuarantorRequestFromJson(json);
}

@freezed
class LoanApplicationResponse with _$LoanApplicationResponse {
  const factory LoanApplicationResponse({
    int? loanId,
    int? accountId,
    String? accountNumber,
    double? principalAmount,
    double? annualInterestRate,
    int? tenureMonths,
    double? emiAmount,
    double? totalPayable,
    double? outstandingBalance,
    double? disbursementCharge,
    LoanStatus? status,
    DateTime? applicationDate,
    DateTime? approvalDate,
    DateTime? disbursementDate,
    DateTime? nextDueDate,
    String? rejectionReason,
    String? disbursementTransactionRef,
    List<GuarantorResponse>? guarantors,
    List<DocumentResponse>? documents,
  }) = _LoanApplicationResponse;

  factory LoanApplicationResponse.fromJson(Map<String, dynamic> json) => _$LoanApplicationResponseFromJson(json);
}

@freezed
class GuarantorResponse with _$GuarantorResponse {
  const factory GuarantorResponse({
    int? id,
    String? name,
    String? phone,
    String? address,
    String? nidNumber,
    String? relation,
    String? photoPath,
  }) = _GuarantorResponse;

  factory GuarantorResponse.fromJson(Map<String, dynamic> json) => _$GuarantorResponseFromJson(json);
}

@freezed
class DocumentResponse with _$DocumentResponse {
  const factory DocumentResponse({
    int? id,
    String? fileName,
    String? originalFileName,
    String? contentType,
    int? fileSize,
  }) = _DocumentResponse;

  factory DocumentResponse.fromJson(Map<String, dynamic> json) => _$DocumentResponseFromJson(json);
}

@freezed
class LoanRepaymentResponse with _$LoanRepaymentResponse {
  const factory LoanRepaymentResponse({
    int? id,
    int? loanId,
    int? installmentNumber,
    DateTime? dueDate,
    double? principalComponent,
    double? interestComponent,
    double? emiAmount,
    double? remainingBalanceAfter,
    RepaymentStatus? status,
    DateTime? paidDate,
    String? transactionRef,
  }) = _LoanRepaymentResponse;

  factory LoanRepaymentResponse.fromJson(Map<String, dynamic> json) => _$LoanRepaymentResponseFromJson(json);
}

@freezed
class LoanScheduleResponse with _$LoanScheduleResponse {
  const factory LoanScheduleResponse({
    int? repaymentId,
    int? installmentNumber,
    DateTime? dueDate,
    double? principalComponent,
    double? interestComponent,
    double? emiAmount,
    double? remainingBalanceAfter,
    RepaymentStatus? status,
    DateTime? paidDate,
    String? transactionRef,
  }) = _LoanScheduleResponse;

  factory LoanScheduleResponse.fromJson(Map<String, dynamic> json) => _$LoanScheduleResponseFromJson(json);
}
